Web Extension: Outline messaging scheme between WebProcess

akutira-umich opened the pull request:
#### 13b3b5ec3ab2205d7b9760824b785db16d8c543f
<pre>
Web Extension: Outline messaging scheme between WebProcess
and UIProcess for bookmarks
<a href="https://rdar.apple.com/154869719">rdar://154869719</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=295339">https://bugs.webkit.org/show_bug.cgi?id=295339</a>

Reviewed by Timothy Hatcher.

Adds message and receiver functions in the WebExtensionContext for connecting the messaging between the Web and UI processes necessary for Bookmarks.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.serialization.in: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm: Added.
(WebKit::WebExtensionContext::isBookmarksMessageAllowed):
(WebKit::WebExtensionContext::bookmarksCreate):
(WebKit::WebExtensionContext::bookmarksGetTree):
(WebKit::WebExtensionContext::bookmarksGetSubTree):
(WebKit::WebExtensionContext::bookmarksGet):
(WebKit::WebExtensionContext::bookmarksGetChildren):
(WebKit::WebExtensionContext::bookmarksGetRecent):
(WebKit::WebExtensionContext::bookmarksSearch):
(WebKit::WebExtensionContext::bookmarksUpdate):
(WebKit::WebExtensionContext::bookmarksMove):
(WebKit::WebExtensionContext::bookmarksRemove):
(WebKit::WebExtensionContext::bookmarksRemoveTree):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/297193@main">https://commits.webkit.org/297193@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d2b91e7a65b0166c23fd2f2bccfdef5530a96ac

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110622 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30281 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20714 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116648 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60889 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30960 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38870 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84120 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113570 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24725 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99607 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64560 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24088 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60443 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94112 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17806 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119438 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37662 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27976 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93082 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38036 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95878 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92905 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37925 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15667 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33639 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17880 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37557 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43028 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37220 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40559 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38928 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

AriYoung00 commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -8015,7 +8018,7 @@
 		CD8252E025D4918400862FD8 /* RemoteRemoteCommandListenerMessageReceiver.cpp */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.cpp; path = RemoteRemoteCommandListenerMessageReceiver.cpp; sourceTree = "<group>"; };
 		CD8252E125D4918500862FD8 /* RemoteRemoteCommandListenerMessages.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = RemoteRemoteCommandListenerMessages.h; sourceTree = "<group>"; };
-		CD95493526159004008372D9 /* .dylib */ = {isa = PBXFileReference; explicitFileType = "compiled.mach-o.dylib"; includeInIndex = 0; path = .dylib; sourceTree = BUILT_PRODUCTS_DIR; };
+		CD95493526159004008372D9 /* libWebKitSwift.dylib */ = {isa = PBXFileReference; explicitFileType = "compiled.mach-o.dylib"; includeInIndex = 0; path = libWebKitSwift.dylib; sourceTree = BUILT_PRODUCTS_DIR; };
```
Looks like this change (and similar) shows up in a few places. Was this intentional?

akutira-umich replied:
No this wasn't. I'm not sure why this changed

AriYoung00 replied:
It would probably be good to revert it in that case

AriYoung00 commented on Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.h:
```diff
@@ -0,0 +1,50 @@
+
+namespace WebKit {
+
+struct WebExtensionBookmarks {
```
Should this be singular?
```suggestion
struct WebExtensionBookmark {
```

akutira-umich replied:
Oh I tried to keep everything consistent by making it plural

AriYoung00 replied:
I think if this only holds the details of one bookmark, it should probably be singular. Also, one of the members of this struct is:
```cpp
std::optional<Vector<Ref<WebExtensionBookmark>>> children;
```
Shouldn't they be the same type?

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -0,0 +1,69 @@
+
+namespace WebKit {
+
+bool isBookmarksAllowed();
```
What is this prototype for?

AriYoung00 commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -2591,7 +2592,6 @@
 		F4E090952D889D6A00322226 /* WKIdentityDocumentPresentmentDelegate.h in Headers */ = {isa = PBXBuildFile; fileRef = F4E090942D889CB800322226 /* WKIdentityDocumentPresentmentDelegate.h */; };
 		F4E090992D88AAC300322226 /* WKIdentityDocumentPresentmentRawRequest.h in Headers */ = {isa = PBXBuildFile; fileRef = F4E090962D88A66F00322226 /* WKIdentityDocumentPresentmentRawRequest.h */; };
 		F4E28A362C923814008120DD /* ScriptTrackingPrivacyFilter.h in Headers */ = {isa = PBXBuildFile; fileRef = F4E28A352C923814008120DD /* ScriptTrackingPrivacyFilter.h */; };
-		F4E44EB72DCD33E800A304C4 /* ISO18013MobileDocumentRequest+Extras.swift in Sources */ = {isa = PBXBuildFile; fileRef = F4E44EB62DCD33E800A304C4 /* ISO18013MobileDocumentRequest+Extras.swift */; };
```
Was this change intentional?

akutira-umich replied:
No, Ill revert these

AriYoung00 commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -21109,7 +21116,6 @@
 				0735F31E2D3760BC0003D029 /* IntelligenceTextEffectChunk.swift in Sources */,
 				0735F31C2D3760AA0003D029 /* IntelligenceTextEffectViewManager.swift in Sources */,
 				E838FCB02DE90BF800703353 /* ISO18013MobileDocumentRequest+Extras.swift in Sources */,
-				A1098E6D2BB6436400449EE0 /* LinearMediaKitExtras.swift in Sources */,
```
Was this change intentional?

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.h:
```diff
@@ -0,0 +1,43 @@
+struct WebExtensionBookmarksParameters {
+    String nodeId;
+    std::optional<String> parentId;
+    size_t index;
```
Per https://github.com/WebKit/WebKit/commit/f3013dee5362c1ffcb855561b13288569a6a21b3.
```suggestion
    uint64_t index;
```

akutira-umich replied:
Based on the PR it sounds like Ill just take out any usage of size_t in the UIProcess and swap it to uint64_t even in the BookmarksCocoa.mm and WebContext.h files

xeenon replied:
Likely.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.serialization.in:
```diff
@@ -0,0 +1,34 @@
+struct WebKit::WebExtensionBookmarksParameters {
+    String nodeId;
+    std::optional<String> parentId;
+    size_t index;
```
```suggestion
    uint64_t index;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -0,0 +1,70 @@
+{
+    // FIXME: fix
+    return false;
+}
```
Would be better at the top of the file.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
```diff
@@ -59,6 +74,8 @@ messages -> WebExtensionContext {
     [Validator=isCookiesMessageAllowed] CookiesRemove(std::optional<PAL::SessionID> sessionID, String name, URL url) -> (Expected<std::optional<WebKit::WebExtensionCookieParameters>, WebKit::WebExtensionError> result)
     [Validator=isCookiesMessageAllowed] CookiesGetAllCookieStores() -> (Expected<HashMap<PAL::SessionID, Vector<WebKit::WebExtensionTabIdentifier>>, WebKit::WebExtensionError> result)
+    
+    
```
Remove.

xeenon reviewed (changes requested):
Almost ready, needs project file changes reverted.

akutira-umich committed:
Web Extension: Outline messaging scheme between WebProcess
and UIProcess for bookmarks
rdar://154869719
https://bugs.webkit.org/show_bug.cgi?id=295339

Reviewed by Timothy Hatcher.

Adds message and receiver functions in the WebExtensionContext for connecting the messaging between the Web and UI processes necessary for Bookmarks.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionBookmarksParameters.serialization.in: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm: Added.
(WebKit::WebExtensionContext::isBookmarksMessageAllowed):
(WebKit::WebExtensionContext::bookmarksCreate):
(WebKit::WebExtensionContext::bookmarksGetTree):
(WebKit::WebExtensionContext::bookmarksGetSubTree):
(WebKit::WebExtensionContext::bookmarksGet):
(WebKit::WebExtensionContext::bookmarksGetChildren):
(WebKit::WebExtensionContext::bookmarksGetRecent):
(WebKit::WebExtensionContext::bookmarksSearch):
(WebKit::WebExtensionContext::bookmarksUpdate):
(WebKit::WebExtensionContext::bookmarksMove):
(WebKit::WebExtensionContext::bookmarksRemove):
(WebKit::WebExtensionContext::bookmarksRemoveTree):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/297193@main
